File: r-package/morie/R/evhpvr.R
# SPDX-License-Identifier: AGPL-3.0-or-later

# Profile negative log-likelihood at (a, b); mu and sigma profiled out.
.ht_nll <- function(xv, yv, a, b) {
  n <- length(xv)
  z <- (yv - a * xv) / xv^b
  m <- sum(z) / n
  s2 <- sum((z - m)^2) / n
  if (s2 <= 0) return(list(f = Inf, mu = m, sd = 0))
  list(f = 0.5 * n * log(s2) + b * sum(log(xv)), mu = m, sd = sqrt(s2))
}

#' Heffernan-Tawn conditional extremes model
#'
#' Formula: Y_j | X = x  =  a_j x + x^(b_j) Z_j   for x > u
#'
#' Fitted by Gaussian likelihood on the exceedance set, with the
#' residual mean and sd profiled out at every (a, b), leaving a
#' two-dimensional bounded search over a in [-1, 1] and b in [0, 1).  A
#' deterministic grid then a golden-section refinement in each
#' coordinate keeps both language arms on the same optimum.
#'
#' @param X An n x 2 matrix; column 1 conditions, on a standard
#'   Laplace or exponential-type scale.
#' @param u Conditioning threshold applied to column 1.
#' @return List with \code{a}, \code{b}, \code{mu_z}, \code{sigma_z},
#'   \code{estimate}, \code{nll}, \code{n_exceed}, \code{n},
#'   \code{method}.
#' @references Heffernan & Tawn (2004), JRSS B 66(3):497-546.
#' @export
Evhpvr <- function(X, u) {
  M <- .s03mat(X)
  n <- nrow(M)
  if (n == 0L) stop("empty input: X has no rows")
  if (ncol(M) != 2L) stop("X must have exactly two columns")
  u <- as.numeric(u)
  keep <- M[, 1] > u
  xv <- M[keep, 1]; yv <- M[keep, 2]
  k <- length(xv)
  if (k < 3L) stop("fewer than three exceedances of u; nothing to fit")
  if (any(xv <= 0))
    stop("the conditioning variable must be positive above u")
  a_lo <- -1; a_hi <- 1; b_lo <- 0; b_hi <- 0.999
  bf <- Inf; a <- 0; b <- 0
  for (i in 0:40) {
    aa <- a_lo + (a_hi - a_lo) * i / 40
    for (j in 0:40) {
      bb <- b_lo + (b_hi - b_lo) * j / 40
      f <- .ht_nll(xv, yv, aa, bb)$f
      if (f < bf) { bf <- f; a <- aa; b <- bb }
    }
  }
  da <- (a_hi - a_lo) / 40
  db <- (b_hi - b_lo) / 40
  gr <- 0.5 * (sqrt(5) - 1)
  for (rep in seq_len(60)) {
    lo <- a - da; hi <- a + da
    for (it in seq_len(40)) {
      cc <- hi - gr * (hi - lo); dd <- lo + gr * (hi - lo)
      if (.ht_nll(xv, yv, cc, b)$f < .ht_nll(xv, yv, dd, b)$f) hi <- dd else lo <- cc
    }
    a <- 0.5 * (lo + hi)
    lo <- max(b - db, b_lo); hi <- min(b + db, b_hi)
    for (it in seq_len(40)) {
      cc <- hi - gr * (hi - lo); dd <- lo + gr * (hi - lo)
      if (.ht_nll(xv, yv, a, cc)$f < .ht_nll(xv, yv, a, dd)$f) hi <- dd else lo <- cc
    }
    b <- 0.5 * (lo + hi)
    da <- da * 0.5
    db <- db * 0.5
  }
  r <- .ht_nll(xv, yv, a, b)
  .t1_result(a = a, b = b, mu_z = r$mu, sigma_z = r$sd, estimate = a,
             nll = r$f, n_exceed = k, n = n,
             method = "Heffernan-Tawn conditional extremes model")
}

File: src/morie/fn/evhpvr.py
# morie.fn -- function file (rootcoder007/morie)
"""Heffernan-Tawn conditional extremes model."""

import math

from . import _s03core as core
from ._richresult import RichResult

__all__ = ["evt_heffernan_tawn"]


def _ht_nll(xv, yv, a, b):
    """Profile negative log-likelihood at (a, b), mu and sigma profiled out."""
    n = len(xv)
    z = [(yv[i] - a * xv[i]) / xv[i] ** b for i in range(n)]
    m = sum(z) / n
    s2 = sum((v - m) ** 2 for v in z) / n
    if s2 <= 0.0:
        return float("inf"), m, 0.0
    return (0.5 * n * math.log(s2) + b * sum(math.log(v) for v in xv),
            m, math.sqrt(s2))


def evt_heffernan_tawn(X, u):
    """
    Heffernan-Tawn conditional extremes model

    Formula: Y_j | X = x  =  a_j x + x^(b_j) Z_j   for x > u

    Fitted by Gaussian likelihood on the exceedance set, with the
    residual mean and sd profiled out at every (a, b), leaving a
    two-dimensional bounded search over a in [-1, 1] and b in [0, 1).
    A deterministic grid then a golden-section refinement in each
    coordinate keeps both language arms on the same optimum.

    Parameters
    ----------
    X : array-like
        n x 2 matrix.  Column 0 is the conditioning variable, already on
        a standard Laplace or exponential-type scale.
    u : float
        Conditioning threshold applied to column 0.

    Returns
    -------
    result : dict
        Keys: a, b, mu_z, sigma_z, estimate (a), n_exceed, n.

    References
    ----------
    Heffernan & Tawn (2004), JRSS B 66(3):497-546.
    """
    M = core.mat(X)
    n = len(M)
    if n == 0:
        raise ValueError("empty input: X has no rows")
    if len(M[0]) != 2:
        raise ValueError("X must have exactly two columns")
    u = float(u)
    xv = [M[i][0] for i in range(n) if M[i][0] > u]
    yv = [M[i][1] for i in range(n) if M[i][0] > u]
    k = len(xv)
    if k < 3:
        raise ValueError("fewer than three exceedances of u; nothing to fit")
    if any(v <= 0.0 for v in xv):
        raise ValueError("the conditioning variable must be positive above u")
    a_lo, a_hi, b_lo, b_hi = -1.0, 1.0, 0.0, 0.999
    best = (float("inf"), 0.0, 0.0)
    for i in range(41):
        a = a_lo + (a_hi - a_lo) * i / 40.0
        for j in range(41):
            b = b_lo + (b_hi - b_lo) * j / 40.0
            f = _ht_nll(xv, yv, a, b)[0]
            if f < best[0]:
                best = (f, a, b)
    a, b = best[1], best[2]
    da = (a_hi - a_lo) / 40.0
    db = (b_hi - b_lo) / 40.0
    gr = 0.5 * (math.sqrt(5.0) - 1.0)
    for _ in range(60):
        lo, hi = a - da, a + da
        for _ in range(40):
            c = hi - gr * (hi - lo)
            d = lo + gr * (hi - lo)
            if _ht_nll(xv, yv, c, b)[0] < _ht_nll(xv, yv, d, b)[0]:
                hi = d
            else:
                lo = c
        a = 0.5 * (lo + hi)
        lo, hi = max(b - db, b_lo), min(b + db, b_hi)
        for _ in range(40):
            c = hi - gr * (hi - lo)
            d = lo + gr * (hi - lo)
            if _ht_nll(xv, yv, a, c)[0] < _ht_nll(xv, yv, a, d)[0]:
                hi = d
            else:
                lo = c
        b = 0.5 * (lo + hi)
        da *= 0.5
        db *= 0.5
    nll, mu, sd = _ht_nll(xv, yv, a, b)
    return RichResult(payload={
        "a": a,
        "b": b,
        "mu_z": mu,
        "sigma_z": sd,
        "estimate": a,
        "nll": nll,
        "n_exceed": k,
        "n": n,
        "method": "Heffernan-Tawn conditional extremes model",
    })


def cheatsheet():
    return "evhpvr: Heffernan-Tawn conditional extremes model"


# compact alias per ledger/NAMING.md
evtheffernantawn = evt_heffernan_tawn
